templates/components/multiagent/typescript/workflow/factory.ts

```diff
@@ -0,0 +1,152 @@
+import {
+  Context,
+  StartEvent,
+  StopEvent,
+  Workflow,
+  WorkflowEvent,
+} from "@llamaindex/core/workflow";
+import { StreamData } from "ai";
+import {
+  BaseToolWithCall,
+  CallbackManager,
+  ChatMemoryBuffer,
+  ChatMessage,
+  EngineResponse,
+  LLM,
+  OpenAIAgent,
+  Settings,
+} from "llamaindex";
+import { AgentInput, FunctionCallingStreamResult } from "./type";
+
+class InputEvent extends WorkflowEvent<{
+  input: ChatMessage[];
+}> {}
+
+export class FunctionCallingAgent extends Workflow {
+  name: string;
+  llm: LLM;
+  memory: ChatMemoryBuffer;
+  tools: BaseToolWithCall[];
+  systemPrompt?: string;
+  writeEvents: boolean;
+  role?: string;
+  callbackManager: CallbackManager;
+  stream: StreamData;
+
+  constructor(options: {
+    name: string;
+    llm?: LLM;
+    chatHistory?: ChatMessage[];
+    tools?: BaseToolWithCall[];
+    systemPrompt?: string;
+    writeEvents?: boolean;
+    role?: string;
+    verbose?: boolean;
+    timeout?: number;
+    stream: StreamData;
+  }) {
+    super({
+      verbose: options?.verbose ?? false,
+      timeout: options?.timeout ?? 360,
+    });
+    this.name = options?.name;
+    this.llm = options.llm ?? Settings.llm;
+    this.memory = new ChatMemoryBuffer({
+      llm: this.llm,
+      chatHistory: options.chatHistory,
+    });
+    this.tools = options?.tools ?? [];
+    this.systemPrompt = options.systemPrompt;
+    this.writeEvents = options?.writeEvents ?? true;
+    this.role = options?.role;
+    this.callbackManager = this.createCallbackManager();
+    this.stream = options.stream;
+
+    // add steps
+    this.addStep(StartEvent<AgentInput>, this.prepareChatHistory, {
+      outputs: InputEvent,
+    });
+    this.addStep(InputEvent, this.handleLLMInput, {
+      outputs: StopEvent,
+    });
+  }
+
+  private get chatHistory() {
+    return this.memory.getAllMessages();
+  }
+
+  private async prepareChatHistory(
+    ctx: Context,
+    ev: StartEvent<AgentInput>,
+  ): Promise<InputEvent> {
+    const { message, streaming } = ev.data.input;
+    ctx.set("streaming", streaming);
+    this.writeEvent(`Start to work on: ${message}`);
+    if (this.systemPrompt) {
+      this.memory.put({ role: "system", content: this.systemPrompt });
+    }
+    this.memory.put({ role: "user", content: message });
+    return new InputEvent({ input: this.chatHistory });
+  }
+
+  private async handleLLMInput(
+    ctx: Context,
+    ev: InputEvent,
+  ): Promise<StopEvent<string | ReadableStream<EngineResponse>>> {
+    const chatEngine = new OpenAIAgent({
```

Comment:
don't create new agent, directly use LLM (let me know if there's something missing in LITS)